### DISCUSSION TITLE
讨论一个技术实现的问题

### DISCUSSION THREAD

**Discussion by ericzhanchina** [cloudwu/skynet#1743] [General]
游戏上线后，难免会因为一些bug造成玩家的数据出现错误，或者因为bug道具某些bug没有能获得，bug修复后，玩家的数据如何修改为正确的数据，漏掉的道具如何补发？首先，肯定要玩家登录游戏，我们才可以根据玩家的数据情况进行分析，得出结论，是否该补发道具，或者修复数据，这就要把代码写入到登录的协议里面去，如果有10个bug，就可能要写10个类似的方法去做这个事情，代码会很难看，而且也会丧失一定的效率。

有没有比较好的设计方式，skynet&lua的新手一枚，请教大佬 :)

**Comment by terry8210** [comment]
通常游戏都有用一些标志位作为一些新手引导和事件记录等等，保留一些ID的范围，然后通过一些标志位去对比系统那些功能需要补发奖励，独立一个服务即可了，然后接入邮件发送奖励。

**Reply by ericzhanchina** [reply]
这些记录是有存储的，你还是需要玩家登录进入游戏之后，才能分析这些数据，然后决定是不是要发邮件补道具，还是直接修改玩家的数据，这只是解了一个bug的，如果这种bug有个10个，那就要写10个类似的分析方法，会不会代码看上去很不好维护呢

**Reply by sniper00** [reply]
要不要补，主要是分析统计日志

**Reply by ericzhanchina** [reply]
分析日志不可能玩家登录进来的时候再去分析日志吧，那样性能就太低了，为何不是玩家登录进来的时候分析玩家的游戏数据决定要不要补发呢，比如如果有一个成就是以宠物技能数为奖励依据的，之前的bug导致玩家可能有了4技能的宠物，但是记录的技能最大数据还是1呢，这种其实就可以玩家进入游戏后检查一遍数据，直接修改最大技能数为4.

**Reply by terry8210** [reply]
独立于游戏其他系统，只是用一个标志位确定对比，然后用邮件进行发奖励，怎么会影响原来的代码，怎么会不好维护呢？

**Reply by ericzhanchina** [reply]
你的意思是你补发道具这种都是通过后台去筛选，然后发送邮件吗？玩家的数据可能是比较复杂的，比如某个时间点之前注册的账号，如果完成了某个支线的玩家，可能需要补某某道具，这种就很难筛选出来的。

**Comment by cloudwu** [comment]
我觉得应该把（一次性）任务发放的道具和从其它途径发放的道具内部区分开。即使玩家来看道具是相同的。

比如新手任务发了一把铁剑，铁剑也可以打怪掉出来。但是内部是区分开两种铁剑的。

对于任务发放的铁剑，即使玩家销毁掉，也只是额外记录这个东西没了，到这个东西是挂在玩家的数据库中的（可以是能数据不加载）。这样，从数据库分析数据，玩家是否获得过任务奖励铁剑一目了然。

**Reply by ericzhanchina** [reply]
云风的意思是离线分析玩家的数据库数据，然后邮件补发吗？这样有较多的注册用户的时候，又有一批玩家受到了bug的影响，那从数据库里面选这些玩家还是有点儿难受的。我是想玩家登录游戏，玩家的数据已经进入了内存的时候，我去分析数据，然后要么修改数据，要么补发邮件道具。

这样做，就要在登录协议里面加上很多if else的判断，感觉很难维护，日久天长，登录的协议效率也降低了。

**Reply by cloudwu** [reply]
一般补发都是通过游戏内邮件的。邮件是可以离线发也可以在线发的。并不需要强制嵌在登录流程去检查。为了玩家方便，倒是可以在登录流程自动检查邮箱看有没有 bug 补偿邮件，自动领取。

**Reply by ericzhanchina** [reply]
例如这种情况，如果有一个成就是以宠物技能数为奖励依据的，之前的bug导致玩家可能有了4技能的宠物，但是记录的技能最大数据还是1呢，就必须玩家进入游戏后检查一遍数据，直接修改最大技能数为4，邮件就无法处理了。而且基本只能登录后靠代码去检查，才能修复这个遗留的问题。这种多起来，就很头疼了，代码被弄的很乱。

**Reply by sniper00** [reply]
> 例如这种情况，如果有一个成就是以宠物技能数为奖励依据的，之前的bug导致玩家可能有了4技能的宠物，但是记录的技能最大数据还是1呢，就必须玩家进入游戏后检查一遍数据，直接修改最大技能数为4，邮件就无法处理了。而且基本只能登录后靠代码去检查，才能修复这个遗留的问题。这种多起来，就很头疼了，代码被弄的很乱。

所以说这种数据分析，主要靠统计日志呀，运营部门也会需要玩家详细的行为日志

**Reply by ericzhanchina** [reply]
日志你说的是skynet.error输出的还是什么日志？这种其实玩法登录进来直接修复他的数据是最合适的，不用跟日志去挂钩，我们用skynet.error记录的日志只是给程序员看的。不知道兄弟说的日志是什么类型的日志。

**Reply by sniper00** [reply]
> 日志你说的是skynet.error输出的还是什么日志？这种其实玩法登录进来直接修复他的数据是最合适的，不用跟日志去挂钩，我们用skynet.error记录的日志只是给程序员看的。不知道兄弟说的日志是什么类型的日志。

并非skynet.error。 玩家行为统计日志，游戏内系统行为日志，如道具增减，活动加分，怪物击杀，副本通关，  一般是有固定格式，方便大数据分析的日志。数据库只是现在时间**点**的状态， 而统计日志是 一本历史书，表示一段时间内的状态变化记录。

**Reply by ericzhanchina** [reply]
这个游戏里面的日志我也做了的，但是像我刚才说的宠物技能这个，如果少算了，有这个bug，只能去分析玩家的数据去修正，比如选择登录的时候。这种积累下来就会感觉有点儿乱糟糟的

**Reply by terry8210** [reply]
或者做数据库查询，查询那些账号有问题，然后补发邮件把，这样就不需要在游戏代码中加入任何逻辑，其实与上面分析日志的思路是一样的。

**Reply by ericzhanchina** [reply]
是，我也是非常不想把代码嵌入到游戏逻辑里面来。补发的这种离线补发倒是还好。

就怕有些数据需要修改的，好像最好是在登录的逻辑里面嵌入这些补丁，修补玩家的数据。

**Comment by hanxi** [comment]
开一个 patch 目录，里面只写一次性的代码，每段代码对应一个 bug 修复，每个老玩家执行一次，然后标记已执行。文件名可以是日期时间开头排序的。

**Reply by ericzhanchina** [reply]
嗯，感觉是这样比较合适，修复玩家的数据，或者通过邮件补发道具都可以在这里做。每次玩家登录进来都要检查一下patch这个模块的方法是否执行过，没执行的都执行一遍。

**Comment by owllyi** [comment]
提供个思路，给需要的玩家发一封邮件，邮件会在用户登陆时自动触发一段修复逻辑并销毁邮件。也可以达到目的。

**Comment by ray2011** [comment]
首先，肯定要玩家登录游戏，我们才可以根据玩家的数据情况进行分析，得出结论，是否该补发道具，或者修复数据，这就要把代码写入到登录的协议里面去

这句话有问题，不需要玩家登录游戏一样可以修复玩家数据，做好 actor 锁就行。第二，是否补发道具，修复数据，简单的可以现场计算，复杂的可以离线计算之后作为配置文件热更上去使用

**Comment by qhf566** [comment]
谁有大厂,近5年内开发完的,国内没有运营过的,成品开发完毕的手机游戏,联系我809996677
